ESP IDF coredump

noorhaq opened the pull request:
Build ID and Coredump needs to be enabled from menuconfig-> components -> Coredump -> enable coredump and save to flash

noorhaq committed:
Added Coredump

noorhaq committed:
Updated idf_component.yml to point to the branch

noorhaq committed:
Added missing build files.

noorhaq committed:
Updated commit hash

noorhaq committed:
Added sdkconfig.defaults

noorhaq committed:
Fixed bug regarding the cloud not receiving logs

noorhaq committed:
Optimized the MQTT sending
Added Coredump bug fix,
Added build ID.

noorhaq committed:
Added build workflow to test examples for any errors.

noorhaq committed:
Added fixes to esp-idf.yml. (#42)

roberthusak commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,103 @@
+
+on:
+  push:
+  pull_request:
```
I recommend keeping either one or the other (probably `push`) so that pushing to a PR doesn't cause launching the action twice.

noorhaq replied:
Done

roberthusak commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,103 @@
+        with:
+          path: |
+            $HOME/esp-idf
+            $HOME/.espressif
```
Doesn't seem to work - see that it's not present https://github.com/spotflow-io/device-sdk/actions/caches

The latest run shows this warning:
```
Warning: Path Validation Error: Path(s) specified in the action for caching do(es) not exist, hence no cache is being saved.
```

noorhaq replied:
Right now removing it will fix the cache in another workflow needs to test it why it's not creating the id in caches.

roberthusak commented on esp_idf/spotflow/examples/Coredump/main/app_main.c:
```diff
@@ -0,0 +1,91 @@
+
+static const char* TAG = "spotflow_testing_coredump";
+
+#define GPIO_INPUT_IO_0     4
```
Is this a button available on ESP32-C6 and ESP32-S3? Or does the developer have to connect their own?

jmasek replied:
@noorhaq cannot we use the boot button?

noorhaq replied:
Moved it to the BOOT button but I still don't wanna connect it to boot.

roberthusak commented on esp_idf/spotflow/examples/Coredump/main/app_main.c:
```diff
@@ -0,0 +1,91 @@
+// ISR handler — must be IRAM safe if registered with ISR service
+static void IRAM_ATTR button_isr_handler(void* arg)
+{
+    // esp_system_abort("Button pressed -> deliberate crash");
```
Remove commented-out code (other occurrences as well).

roberthusak commented on esp_idf/spotflow/examples/Coredump/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,23 @@
+CONFIG_ESP_COREDUMP_ENABLE_TO_FLASH=y
+# CONFIG_ESP_COREDUMP_ENABLE_TO_UART is not set
+# CONFIG_ESP_COREDUMP_ENABLE_TO_NONE is not set
+# CONFIG_ESP_COREDUMP_DATA_FORMAT_BIN is not set
```
Do we need to keep the unused options in the comments?

noorhaq replied:
I have captured the default ones like the complete block because if any 2 are active at the same time it will cause issues.

roberthusak commented on esp_idf/spotflow/include/buildid/spotflow_build_id.h:
```diff
@@ -0,0 +1,16 @@
+
+int spotflow_build_id_get(const uint8_t** build_id, uint16_t* build_id_len);
+void spotflow_build_id_print(void);
+#ifdef __cplusplus
```
Nitpick: Add a newline before `#ifdef` for consistency (in other header files as well).

roberthusak commented on esp_idf/spotflow/include/coredump/spotflow_coredump_queue.h:
```diff
@@ -0,0 +1,19 @@
+int8_t queue_coredump_push(uint8_t* msg, size_t len);
+bool queue_coredump_read(queue_msg_t* out);
+void queue_coredump_free(queue_msg_t* msg);
+void queue_coredump_init(void);
```
Shouldn't all our non-static symbols be prefixed with `spotflow_` so that they don't collide with user symbols?

roberthusak commented on esp_idf/spotflow/src/coredump/spotflow_coredump_cbor.c:
```diff
@@ -0,0 +1,203 @@
+/* Should be approximately 47 bytes (including build ID), putting 64 to be safe */
+#define COREDUMPS_OVERHEAD 64
+
+uint8_t buffer[CONFIG_SPOTFLOW_COREDUMPS_CHUNK_SIZE + COREDUMPS_OVERHEAD];
```
Unused. In Zephyr, it's used to construct the CBOR message and determine its length, then a buffer of this length is allocated and the CBOR message is copied to it. But maybe your approach here is easier - although the result might occupy a slightly larger memory chunk than it needs, the buffer doesn't have to be copied. So maybe just remove `buffer`. @jmasek What do you think?

noorhaq replied:
Yep this was a oversight. I had planned to use stack then moved to using the heap. For this. So, i am gonna remove it.

roberthusak commented on esp_idf/spotflow/src/coredump/spotflow_coredump_cbor.c:
```diff
@@ -0,0 +1,203 @@
+	// Encode key-value pairs one by one
+	err = cbor_encode_uint(&map_encoder, KEY_MESSAGE_TYPE);
+	if (err != CborNoError)
+		goto fail;
```
Add braces `{}` for consistency.

roberthusak commented on esp_idf/spotflow/src/coredump/spotflow_coredump_cbor.c:
```diff
@@ -0,0 +1,203 @@
+ * @param buf
+ * @param len
+ */
+static void print_cbor_hex(const uint8_t* buf, size_t len)
```
I'd remove this unused (or at least with the commented-out use) function. I imagine that it might have been useful during the development, but I'm not sure it's worth keeping there. @jmasek What do you think?

The same holds for `print_cbor_hex` in `spotflow_log_cbor.h`. If we want to keep those here, shouldn't we maybe introduce an additional Kconfig option instead of having the function call commented out?

noorhaq replied:
I think it's worth keeping because if we want to see if anything went wrong with any new boards or configs we can print the raw frames on the cbor side to verify at what point it got corrupted. I will comment it out.

roberthusak replied:
I'd prefer to have an `#ifdef` with a Kconfig option instead of commenting the code. I think we should avoid commented code in `main` if possible (it' OK for experimenting in PRs though).

roberthusak commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -65,8 +65,8 @@ menu "Spotflow"
+    config SPOTFLOW_COREDUMPS_CHUNK_SIZE
+        int 
         default SPOTFLOW_CBOR_LOG_MAX_LEN
 endmenu
```
Shouldn't be the sending of core dumps using Spotflow configurable in Kconfig? It's opt-in in the Zephyr module.

noorhaq replied:
This one is not exposed. It just makes sure the COREDUMPS_CHUNK_SIZE is equal to SPOTFLOW_CBOR_MAX_LEN to make sure the mqtt buffer can handle them as MQTT buffer is being checked against ```SPOTFLOW_CBOR_LOG_MAX_LEN```

roberthusak commented on esp_idf/spotflow/include/logging/spotflow_log_queue.h:
```diff
@@ -1,14 +1,22 @@
-void queue_free(queue_msg_t *msg);
+void queue_push(uint8_t* msg, size_t len);
+bool queue_read(queue_msg_t* out);
+void queue_free(queue_msg_t* msg);
```
Shouldn't these functions be renamed to `queue_log_*`? Or rather `spotflow_queue_log_*`?

roberthusak commented on esp_idf/spotflow/spotflow.c:
```diff
@@ -16,19 +23,23 @@ vprintf_like_t original_vprintf = NULL;
  */
 void Spotflow_Todo(void)
```
Is this function still needed?

noorhaq replied:
Yep for future add ons and stuff.

jmasek replied:
Can you explain it a bit more? Why won't you add it when needed in the future?

noorhaq replied:
I have removed it. Like I thought we can have it for consistency. Because we will always have something in todo.

roberthusak commented on esp_idf/spotflow/src/coredump/spotflow_coredump.c:
```diff
@@ -0,0 +1,214 @@
+static coredump_info_t coredump_info = { 0 };
+
+/**
+ * @brief If coredump Partion is available return true otherwise false
```
It also checks whether the coredump is actually stored in the partition, right?

noorhaq replied:
Yes, first it checks if the partition exists if yes then it checks whether a valid coredump image exists in the partition or not.

roberthusak commented on esp_idf/spotflow/src/coredump/spotflow_coredump.c:
```diff
@@ -0,0 +1,214 @@
+			else
+			{
+				vTaskDelay(pdMS_TO_TICKS(50)); //50ms wait //To avoid watchdog trigger
+			}
```
`if` and `else` branches look the same. Btw. why the watchdog could be triggered? Isn't the thread preemptive?

noorhaq replied:
Yep, kind of a mismatch at my part. It is I was adding a delay because it was retrying and getting a lot of messages on the terminal.

roberthusak reviewed (commented):
Looks good, I added mostly minor and stylistic comments.

jmasek commented on esp_idf/spotflow/CMakeLists.txt:
```diff
@@ -7,8 +8,41 @@ set(SRCS
+endif()
+
+# ---- Optional coredump sources ----
+if(CONFIG_ESP_COREDUMP_ENABLE)
```
I do not think this is the correct approach. If the user wants to use our coredump backend -> we should turn on all the necessary ESP components. Not the other way around - if the user activates ESP coredumps, as a side effect, we activate our backend.

noorhaq replied:
Ahh kind of confusion here as when we enable the coredump from the menuconfig like using our spotflow configurations it will activate the coredump too. But due to any reason if the coredump is not enabled. And we don't add a check here all the files inside will fail. So that's why I added this one here and I think we shouldn't change it.

jmasek replied:
so shouldn't this be `if(SPOTFLOW_COREDUMP_BACKEND)` ?

noorhaq replied:
this is like a dummy check. because we are enabling this config in the Kconfig like the SPOTFLOW_COREDUMP_BACKEND is a wrapper for this and some more. If this fails then everything will fail. We can add duplicate checks.

jmasek replied:
why duplicate check - if `SPOTFLOW_COREDUMP_BACKEND` is not true, there is no reason add these source file to build, right? or do I miss something?

noorhaq replied:
Well yes but we can add like 2 checks because if the spotflow coredump is enabled but at the same time in some config or component they have disabled the ESP_Coredump then we will have linking issues/errors. 
So, in the new Kconfig one we can make it more options and add this in wraps.

jmasek replied:
If someone disables somewhere else `CONFIG_ESP_COREDUMP_ENABLE` -> the solution should be triggering error/warning, not silently omitting our source code!
But disabling `SPOTFLOW_COREDUMP_BACKEND` and therefore omitting our source code is totally fine setup

noorhaq replied:
Exactly that's what being added in the Kconfigs and Cmake right now. 
So these are double checked and all the configs we require are added.

jmasek replied:
ok, so it will be here or in a separate PR?

noorhaq replied:
Seperate PR. Like the next PR has configs options and a MQTT 5 exposed options. Then on a call we can create the Kconfig. As it is minimal just structural changes.

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -65,8 +65,8 @@ menu "Spotflow"
-    
-    config MQTT_BUFFER_SIZE
-        int
+
```
See comment in the cmakelists.txt
Same applies here 

> If the user wants to use our coredump backend -> we should turn on all the necessary ESP components. Not the other way around - if the user activates ESP coredumps, as a side effect, we activate our backend.
I suggest using something like if possible
```
menu "SPOTFLOW Core Dump Settings"

    # Enable ESP-IDF core dump
    config SPOTFLOW_COREDUMP_BACKEND
        bool "Enable automatic core dump settings"
        default n

    if SPOTFLOW_COREDUMP_BACKEND

        # Default destination = Flash
		config ESP_COREDUMP_TO_FLASH
				  bool
				  default y
				
		# Default full .bss/.data + heap dump
		config ESP_COREDUMP_CAPTURE_HEAP_AND_STACK
				  bool
				  default y


    endif

endmenu
```

noorhaq replied:
Done.

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -65,8 +65,8 @@ menu "Spotflow"
-    config MQTT_BUFFER_SIZE
-        int
+
+    config SPOTFLOW_COREDUMPS_CHUNK_SIZE
```
this config options should be available only if spotflow_coredump is enabled

noorhaq replied:
Done.

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -65,8 +65,8 @@ menu "Spotflow"
-        int
+
+    config SPOTFLOW_COREDUMPS_CHUNK_SIZE
+        int 
```
A couple more questions regarding the config.
* Is it possible to use only coredump functionality without logs? 
* Shouldn't be the config options more organized so log options are available only when log backend enabled? same for coredump?

noorhaq replied:
I will create a seperate PR on this. As this will cause conflict issues and require testing with each component separately enabled. Plus configs on examples will change too.

jmasek replied:
ok, added to milestone

jmasek commented on esp_idf/spotflow/examples/coredump/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+include($ENV{IDF_PATH}/tools/cmake/project.cmake)
+# "Trim" the build. Include the minimal set of components, main, and anything it depends on.
+idf_build_set_property(MINIMAL_BUILD ON)
```
is there a reason why we are using this? does any functionality depends on it?

noorhaq replied:
Well it is providing 2 purposes.
One faster builds and smaller size as only components defined in dependencies are build rest are ignored.
Second it makes sure we have included all the components as requirements which are required by our components
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/build-system.html (Here dp a Ctrl+F search on minimal_build)

jmasek replied:
I see, but my point was that the default is `Off` so I was hesitant to turn it on by default. But I see it is used in esp-idf examples so keep it that way.

jmasek commented on esp_idf/spotflow/src/buildid/spotflow_build_id.c:
```diff
@@ -0,0 +1,47 @@
+ */
+int spotflow_build_id_get(const uint8_t** build_id, uint16_t* build_id_len)
+{
+	 const esp_app_desc_t* app_desc = esp_app_get_description();
```
I see you are using a default esp idf build id! nice. Maybe, do you how exactly it works, what parts are included/omitted in the build id?

noorhaq replied:
Yes more details can be found here
 https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/app_image_format.html

jmasek replied:
there are limited information on the page, However, I checked the python and it seems that it does hash from whole file. It is not ideal. Let's keep it that way and we can discuss it in next milestone.

jmasek commented on esp_idf/spotflow/examples/coredump/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# The following four lines of boilerplate have to be in your project's CMakeLists
+# in this exact order for cmake to work correctly
+cmake_minimum_required(VERSION 3.16)
```
is there a reason why there is capital C in `Coredumps` folder?

noorhaq replied:
Their are examples on the espcoredump components with same name. Either way we can rename it to spotflow_coredump?

jmasek replied:
yes, `spotflow_coredump` and `spotflow_logs` sounds good. btw. where does the conflict comes from? I would expect that the examples are there only for reference and if you are not using both of them simultaneously, it is not problem, or do I miss something?

noorhaq replied:
Well not their but like say we publish to esp component registery and their it will be a conflict as 2 examples with same name like spotflow/coredump
And espcoredump/coredump. 
But as I write it with the prefix it won't cause confusion.

jmasek commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,84 @@
+name: Build ESP-IDF Project
```
Next time, could you please keep features separated in separate PRs so we can merge PRs quicker and comments do not mix together?

jmasek commented on esp_idf/spotflow/examples/Coredump/main/app_main.c:
```diff
@@ -0,0 +1,91 @@
+	ESP_ERROR_CHECK(esp_netif_init());
+    ESP_ERROR_CHECK(esp_event_loop_create_default());
+
+    // Register our custom event handler
```
please add comment what is the event_handler used for

noorhaq replied:
Done. Here it was just used for the button event.

jmasek commented on esp_idf/spotflow/examples/Coredump/main/app_main.c:
```diff
@@ -0,0 +1,91 @@
+
+	spotflow_init();
+	while (1) {
+		// if(atomic_load(&mqtt_connected))
```
commented code cleanup - please uncomment or remove. Or add a description if it can be used for a specific purpose.

jmasek commented on esp_idf/spotflow/examples/Coredump/main/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+idf_component_register(SRCS "app_main.c"
+                    PRIV_REQUIRES mqtt esp_partition nvs_flash esp_netif app_update esp_driver_gpio
```
why we need app_update? because of buildId?

noorhaq replied:
Yes and No. TBH for the build id we require the esp_app_format which is added in the ```spotflow/Cmakelist.txt```
But due to some reason without the app_update component it refuses to recognize and compile. ```app_update``` like I tried to debug but had to give up.

jmasek commented on esp_idf/spotflow/examples/Coredump/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,23 @@
+CONFIG_EXAMPLE_WIFI_SSID=""
+CONFIG_EXAMPLE_WIFI_PASSWORD=""
```
Shouln't be these options rather in sdkonfig.defaults than in KConfig.projbuild?

jmasek commented on esp_idf/spotflow/spotflow.c:
```diff
@@ -16,19 +23,23 @@ vprintf_like_t original_vprintf = NULL;
+	Spotflow_Todo(); //Checking for unused set Configs.
+	queue_init(); //Initilize the queue
+	mqtt_app_start(); // Calling the mqtt_start from the init function.
+#ifdef CONFIG_ESP_COREDUMP_ENABLE
```
Please prefer using our custom KConfig option that the default esp one

jmasek commented on esp_idf/spotflow/examples/coredump/main/app_main.c:
```diff
@@ -0,0 +1,91 @@
+    ESP_ERROR_CHECK(esp_event_handler_register(SPOTFLOW_EVENTS, ESP_EVENT_ANY_ID,
+                                               &spotflow_event_handler, NULL));
+
+	 gpio_config_t btn_conf = {
```
the setup is getting quite bulky - could you split the code in the main into multiple functions? Like 
```
setup_and_checks(){
ESP_LOGI(TAG, "[APP] Startup..");
	ESP_LOGI(TAG, "[APP] Free memory: %" PRIu32 " bytes", esp_get_free_heap_size());
	ESP_LOGI(TAG, "[APP] IDF version: %s", esp_get_idf_version());

	ESP_ERROR_CHECK(nvs_flash_init());
	ESP_ERROR_CHECK(esp_netif_init());
    ESP_ERROR_CHECK(esp_event_loop_create_default());
...
}
button_setup(){
  ....
}
```

noorhaq replied:
It's a example. I have moved the buttons setup to another function but for rest we can keep it as all ESP_IDF based examples use these setups and checks. Like half of this is optional.

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -26,131 +33,167 @@ extern const uint8_t mqtt_spotflow_io_pem_start[]   asm("_binary_x1_root_pem_sta
  * @param pvParameters 
  */
-void mqtt_publish(void *pvParameters)
+void mqtt_publish(void* pvParameters)
```
It seems that `pvParameters` are not used - do we expect that we will be passing some in the future? does it really make sense to have it there? or is it some convention to rather have it than changing interface?

noorhaq replied:
It is a convention and the function the xTaskCreate and other Task APIs expect it. Even when we are not passing any like here we are passing NULL.

noorhaq committed:
Addressed PR change requests (#43)

- Modified the esp-idf.yaml job to only trigger on push and fixed the
cache issue.
- Modified the Kconfig.projbuild
- Modified the queue functions.
- Renamed the coredump example. 
- Modified the coredump example to use boot button.

jmasek commented on esp_idf/spotflow/src/coredump/spotflow_coredump_queue.c:
```diff
@@ -0,0 +1,90 @@
+	// Try to enqueue
+	if (xQueueSend(queue_handle, &qmsg, 0) != pdPASS) {
+        free(qmsg.ptr);
+        return -1; //Could not add wait for it to be freed.
```
mixing tabs and spaces. Can you unify it, please?

noorhaq replied:
Done

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -26,131 +33,157 @@ extern const uint8_t mqtt_spotflow_io_pem_start[]   asm("_binary_x1_root_pem_sta
-                   }
+	queue_msg_t msg;
+
+    while (1) {
```
could you please split this main loop to logical parts (functions) - it got quite long.

noorhaq replied:
Yep in the configs one I have. So the next PR will have it addressed. Can we keep it as it is in this one. Otherwise I will have to spend time on merge conflicts and such as in the next one this file has many changes.

jmasek replied:
ok

noorhaq commented:
I have addressed most of the changes. let me know if any other changes are required.

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -26,131 +33,157 @@ extern const uint8_t mqtt_spotflow_io_pem_start[]   asm("_binary_x1_root_pem_sta
+                    } else {
+                        SPOTFLOW_LOG("Coredump message sent successfully. Freeing queue entry.");
+                        spotflow_queue_free(&msg);
+						vTaskDelay(pdMS_TO_TICKS(10)); //Give CPU few ticks
```
What is the purpose of this?

noorhaq replied:
So, when the Coredumps messages are sent successfully like all we free the queue allocated to it and free all the memory plus delete the current coredump we have. Because the queue memory for the coredump won't be used until a new one is generated and the device is restarted.

noorhaq committed:
Removed whitespace and the labels_encoder from spotflow_log_cbor.c

jmasek commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,88 @@
+          source $HOME/esp-idf/export.sh
+          cd "${{ matrix.example }}"
+          idf.py set-target ${{ matrix.target }}
+          idf.py build || true  # continue-on-error
```
is this intentional? you won't know if pipeline failed if you do not check the status directly.

jmasek commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,88 @@
+          source $HOME/esp-idf/export.sh
+          cd "${{ matrix.example }}"
+          idf.py set-target ${{ matrix.target }}
+          idf.py build || true  # continue-on-error
```
Could you use `CFLAGS="-Werror"` so we know that the build was successful without warnings?

noorhaq replied:
I will add this shortly

jmasek commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,88 @@
+          echo "✅ Found example path: ${{ matrix.example }}"
+
+
+      - name: Install ESP-IDF v5.5.1
```
did you consider using official esp-idf docker image so we do not need to install the toolchain everytime? https://hub.docker.com/layers/espressif/idf/release-v5.5/images/sha256-1948c18c8accdd7b94f7e106ab3908a0dfe1898a7acae159ee92e85a70fb6d06

noorhaq replied:
I did but on the repo their was some restrictions on using third party runs in the organisational repo. 
So had to go with this custom one. I will make it better like adding caching so it won't build it and download every time just once and have it cached to use for future runs.

noorhaq replied:
https://github.com/espressif/idf-ci I am talking about this one

jmasek replied:
what do you mean? I mean using the official docker image that is under Apache Licence 2.0 instead of `ubuntu-latest`

noorhaq replied:
Yep now using that one now in the updated job I tried using the official idf-ci image before but that failed, so i just switched to installing.

jmasek commented on esp_idf/spotflow/examples/coredump/main/app_main.c:
```diff
@@ -0,0 +1,117 @@
+
+static const char* TAG = "spotflow_testing_coredump";
+
+#if CONFIG_IDF_TARGET_ESP32
+    #define GPIO_INPUT_IO_0 0     // BOOT button on ESP32
+#elif CONFIG_IDF_TARGET_ESP32C3
+    #define GPIO_INPUT_IO_0 9     // BOOT button on many C3 devkits
+#elif CONFIG_IDF_TARGET_ESP32C6
+    #define GPIO_INPUT_IO_0 9     // C6 default BOOT button
+#elif CONFIG_IDF_TARGET_ESP32S3
+    #define GPIO_INPUT_IO_0 0     // S3 devkit button (varies)
+#elif CONFIG_IDF_TARGET_ESP32H2
+    #define GPIO_INPUT_IO_0 9     // Example; depends on board
+#else
+    #error "Unsupported target. Please define GPIO_INPUT_IO_0."
+#endif
```
nice, thanks!

jmasek commented on .github/workflows/esp-idf.yml:
```diff
@@ -0,0 +1,88 @@
+          SAFE_NAME="firmware-${{ matrix.target }}-${EXAMPLE_NAME}"
+          echo "name=$SAFE_NAME" >> "$GITHUB_OUTPUT"
+
+      - name: Upload build artifacts
```
Does it make sense to keep built artifacts? Is there any use for them?

noorhaq replied:
Well kind of yes and no. I am looking into running it as qemu in the runs as well like it will test the scripts to see if the data is being sent as well. So with those the artefacts will be necessary and will be deleted afterwards.

jmasek replied:
ok, that would be great, not only build but also run in qemu!
I just checked the esp qemu fork, it looks nice. Keep it there. Should I create an extra milestone for it?

noorhaq replied:
Not now. I still have to check if github actions will support for qemu that plus we will need some checks from cloud side like a exposed something to see the data is reached otherwise it would be a dummy check here that the message is sent which will on QOS 0. But has that message been received or not will not be verified.

jmasek commented on esp_idf/spotflow/examples/coredump/main/idf_component.yml:
```diff
@@ -0,0 +1,6 @@
+  protocol_examples_common:
+    path: ${IDF_PATH}/examples/common_components/protocol_examples_common
+  spotflow:
+    override_path: '../../../'
```
should this be changed when released? Btw. our sdk is already in esp-idf component registry?

noorhaq replied:
(I have got the name and got my account approved it requires a few more modifications like small things to make it fully compatible with the component parser and the rest to see and resolve it) in that I was looking and all the examples had it like this like the override_path works a little differently than the path.
And most components had these in examples like this was the recommended way for examples embedded inside the components.
Like separate readmes for each example is required as well.

jmasek replied:
ok, cool.
So it means the default usage of the example is building it and running it from within the library, right? not copying it out to your own folder/project.

noorhaq replied:
yes.

noorhaq committed:
Added suggestions to the esp-idf.yml and fixed some warnings in Kconfig.

noorhaq committed:
Fixed megalinter error.

noorhaq committed:
Added new line for megalinter.

jmasek commented on esp_idf/spotflow/src/coredump/spotflow_coredump.c:
```diff
@@ -0,0 +1,207 @@
+	// Retrieve the coredump image address and size
+	esp_err_t err = esp_core_dump_image_get(&coredump_addr, &coredump_size);
+	if (err != ESP_OK) {
+		SPOTFLOW_LOG( "No Coredump found.");
```
There can be various errors, but not all of them mean that the coredump was not found, right?

noorhaq replied:
Yes but we can print the exact error as well. I just didn't see the need for it here. Like other places. Because if it's not able to read the coredump for any reason it won't be able to send. And this coredump is handled by esp idf own execution so we don't have much control over it.

jmasek replied:
wouldn't be better to print at least the err code? or something that suggest - that maybe we were not able to read coredump or maybe it is missing, or is corrupted, or something else?

noorhaq replied:
As you saw. If this fails esp_coredump itself does print this stuff. Like what actually went wrong say size too small and other stuff so it would just be duplicates at this point.

jmasek replied:
Error `No Coredump found.` is misleading - if you want to print something (that is definitely reasonable), print `unable to get dump image` no misleading information.

noorhaq replied:
Yeah we can change this to something like this. 
Like the thing here is the first one checks if the partition is present or not.
Then it checks for CRC because there will be data inside whether that data is actually a valid coredump or not. 

So, I will change this.

jmasek commented on esp_idf/spotflow/src/coredump/spotflow_coredump.c:
```diff
@@ -0,0 +1,207 @@
+		const uint8_t* build_id = NULL;
+		uint16_t build_id_len = 0;
+
+#ifdef CONFIG_SPOTFLOW_GENERATE_BUILD_ID
```
Maybe `CONFIG_SPOTFLOW_USE_BUILD_ID` would be better - we are not generating it anymore, more like using a generated one.

jmasek commented on esp_idf/spotflow/src/coredump/spotflow_coredump.c:
```diff
@@ -0,0 +1,207 @@
+
+		// keep retrying to push data until the mqtt function is able to read and clear it
+		do {
+			rc = spotflow_queue_coredump_push(cbor_data, cbor_data_len);
```
why you are using taskDelay here, when you can use `xQueueSend` with xTicksToWait? is it some kind of optimization?

noorhaq replied:
Nope this will get optimized out in the new mqtt.c as that one is not relying on this. Queue push data it creates a task notify which will trigger the mqtt publish. and same for others. Instead of repeating the same until the mqtt send is able to publish.

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -26,131 +33,147 @@ extern const uint8_t mqtt_spotflow_io_pem_start[]   asm("_binary_x1_root_pem_sta
+                    }
+                }
+                else {
+                    vTaskDelay(pdMS_TO_TICKS(50)); // Sleep if queues empty
```
is there a reason why you are using vTaskDelay rather than xTicksToWait in method `xQueueReceive`

noorhaq replied:
These are just temporary as I mentioned before like the mqtt.c file or net files itself has a whole revamp in the new PR with configs these were basically placeholders so I can have the logic and other things stabilised.

Most of this will be completed in the configs PR and after that their will be a small PR on cleanup and optimisation to reduce memory footprints test with different levels of configuration like build id on and off, logs off coredumps on etc.

jmasek replied:
ok

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -1,5 +1,33 @@
+                default ESP_COREDUMP_DATA_FORMAT_ELF
+            
+            config ESP_COREDUMP_CAPTURE_DRAM
+                default y
```
why you are not using select/imply?

noorhaq replied:
I tried that but the Cmake was giving error as this is choice parameters so select and imply weren't setting the values.

noorhaq replied:
And as discussed previously all of these will get their respective SPOTFLOW_Wrapper
Like here for instance SPOTFLOW_COREDUMP_CAPTURE_DRAM. default y but users will be able to modify it in menuconfigs.

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -1,5 +1,33 @@
+            endchoice
+            
+            config ESP_COREDUMP_ENABLE
+                default ESP_COREDUMP_DATA_FORMAT_ELF
```
I do not understand this - how coredump enable is connected to `ESP_COREDUMP_DATA_FORMAT` . I would expect you to select/imply esp_coredump_enable

noorhaq replied:
Oh yeah this one was my mistake fixed it. This one should have had select.

noorhaq committed:
Added custom partition table.

noorhaq committed:
Addressed the PR comments.

noorhaq committed:
Updated Kconfig

jmasek committed:
Merge branch 'main' into feature/esp_idf_coredump

noorhaq committed:
Updated Changelog.

noorhaq committed:
Removed trailing whitespace for Megalinter.
